src/Blocks/BlockCauldron.h

```diff
@@ -27,17 +27,37 @@ class cBlockCauldronHandler :
 	virtual bool OnUse(cChunkInterface & a_ChunkInterface, cWorldInterface & a_WorldInterface, cPlayer & a_Player, int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace, int a_CursorX, int a_CursorY, int a_CursorZ) override
 	{
 		NIBBLETYPE Meta = a_ChunkInterface.GetBlockMeta({a_BlockX, a_BlockY, a_BlockZ});
-		switch (a_Player.GetEquippedItem().m_ItemType)
+		auto EquippedItem = a_Player.GetEquippedItem();
+		switch (EquippedItem.m_ItemType)
 		{
+			case E_ITEM_BUCKET:
+			{
+				if (Meta == 3)
+				{
+					a_ChunkInterface.SetBlockMeta(a_BlockX, a_BlockY, a_BlockZ, 0);
+					// Give new bucket, filled with fluid when the gamemode is not creative:
+					if (!a_Player.IsGameModeCreative())
+					{
+						auto & Inventory = a_Player.GetInventory();
+						auto NewItem = cItem(E_ITEM_WATER_BUCKET);
+						if (Inventory.ReplaceOneEquippedItem(NewItem) == 0)
+						{
+							// The bucket didn't fit, toss it as a pickup:
+							a_Player.TossPickup(NewItem);
+						}
+					}
+				}
+				break;
+			}
 			case E_ITEM_WATER_BUCKET:
 			{
 				if (Meta < 3)
 				{
 					a_ChunkInterface.SetBlockMeta(a_BlockX, a_BlockY, a_BlockZ, 3);
 					if (!a_Player.IsGameModeCreative())
 					{
-						a_Player.GetInventory().RemoveOneEquippedItem();
-						a_Player.GetInventory().AddItem(cItem(E_ITEM_BUCKET));
+						// Remove water bucket, add empty bucket.
+						a_Player.GetInventory().SetEquippedItem(cItem(E_ITEM_BUCKET));
```

Comment:
This should use the `ReplaceOneEquippedItem` function as well, in case plugins in the future decide that water buckets are stackable.